lib/sse_test_server/sse_server.ex

```diff
@@ -0,0 +1,134 @@
+defmodule SSETestServer.SSEServer do
+  @moduledoc """
+  An HTTP server for testing SSE clients.
+
+  TODO: Document properly.
+  """
+  use GenServer
+
+  alias SSETestServer.SSEHandler
+
+  defmodule State do
+    defstruct listener: nil, port: nil, sse_endpoints: %{}
+  end
+
+  defmodule SSEEndpoint do
+    @enforce_keys [:path]
+    defstruct path: nil, handler_state: nil, streams: []
+
+    def new(path, handler_opts \\ []) do
+      handler_state = %SSEHandler.State{path: path, sse_server: self()}
+      %__MODULE__{
+        path: path,
+        handler_state: Map.merge(handler_state, Map.new(handler_opts)),
+      }
+    end
+
+    def to_handler(endpoint),
+      do: {endpoint.path, SSEHandler, endpoint.handler_state}
+  end
+
+  ## Client API
+
+  def start_link(args, opts \\ [name: :sse_test_server]) do
+    GenServer.start_link(__MODULE__, args, opts)
+  end
+  def port(sse \\ :sse_test_server), do: GenServer.call(sse, :port)
+  def base_url(sse \\ :sse_test_server), do: "http://localhost:#{port(sse)}"
+
+  def add_endpoint(sse \\ :sse_test_server, path, handler_opts \\ []),
+    do: GenServer.call(sse, {:add_endpoint, path, handler_opts})
+
+  def event(sse \\ :sse_test_server, path, event, data),
+    do: GenServer.call(sse, {:event, path, event, data})
+
+  def keepalive(sse \\ :sse_test_server, path),
+    do: GenServer.call(sse, {:keepalive, path})
+
+  def end_stream(sse \\ :sse_test_server, path),
+    do: GenServer.call(sse, {:end_stream, path})
+
+  ## Internal client API
+
+  def sse_stream(sse, path, pid),
+    do: GenServer.call(sse, {:sse_stream, path, pid})
+
+  ## Callbacks
+
+  def init(args) do
+    # Trap exits so terminate/2 gets called reliably.
+    Process.flag(:trap_exit, true)
+    listener_ref = make_ref()
+    ranch_args = args |> Enum.filter(fn {k, _} -> k in [:port] end)
```

Comment:
https://ninenines.eu/docs/en/ranch/1.4/manual/ranch_tcp/ has the complete list.

Reply:
Ah, right. The filter was just the easiest way I found to write "if `:port` is in `args`, put it in `ranch_args` otherwise don't" and it also makes it easier to add more stuff later if we decide we want to. (I suppose I could have written the filter function as `fn {:port, _} -> true; _ -> false end` but that's less extensible and not really any shorter.)